CONTRIBUTING.md

```diff
@@ -4,4 +4,415 @@ Thanks for considering contributing to Composer!
+        will install everything in `install_requires`, along with `monai` and `scikit-learn`.
+    1.  It must also be specified in the `run_constrained` and the `test.requires` section.
+    1.  The import must be conditionally imported in the code. For example:
+
+        ```python
+        def unet():
+            try:
+                import monai
+            except ImportError as e:
+                raise ImportError(textwrap.dedent("""\
+                    Composer was installed without unet support. To use unet with Composer, run: `pip install mosaicml
+                    [unet]` if using pip or `conda install -c conda-forge monai` if using Anaconda""") from e
+        ```
+
+        This style allows users to perform minimal install of Composer without triggering `ImportError`s if
+        an optional dependency is missing.
+
+    1.  If the dependency is core to Composer, add the dependency to the `install_requires` section of
+        [setup.py](./setup.py) and the `requirements.run` section of [meta.yaml](./meta.yaml).
+
+### 6.2 Use of `__all__`
+
+All public classes must define `__all__` to be the list of members that should be re-exported.
+The variable is necessary to limit what `from XXX import *` imports, and so the documentation build
+only documents exported members (and not unrelated re-imports).
+
+For example, from [composer/callbacks/memory_monitor.py](composer/callbacks/memory_monitor.py)
+
+```python
+"""Log memory usage during training."""
+import logging
+from typing import Dict, Union
+
+import torch.cuda
+
+from composer.core import Logger, State
+from composer.core.callback import Callback
+
+log = logging.getLogger(__name__)
+
+__all__ = ["MemoryMonitor"]  # export only the MemoryMonitor, not other imports like `Logger`, `State`, or `Callback`
+
+
+class MemoryMonitor(Callback):
+    ...
+```
+
+
+### 6.3 `__init__.py`
+
+All public classes and functions should be added to the module's `__init__.py`.
+
+```python
+from composer.path.to.module.file import MyClass as MyClass
+from composer.path.to.module.file import my_func as my_func
+```
+
+If a file only contains public functions, then the following is also acceptable:
+
+```python
+from composer.path.to.module import my_file as my_file
+```
+
+
+## 7. Documentation
+
+Composer uses [Google Style Docstrings](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).
+
+The following guidelines apply to documentation.
+1.  Each function that needs a docstring must have its input arguments and return statement (if not None) annotated.
+1.  Each argument annotation should include the type. If the argument has a default value, the type annotation should
+    specify "optional", and the docstring should say the default value. There should only be one line for the "Returns:"
+    section, even if a tuple is returned. For example:
+
+    ```python
+
+    def foo(bar: int):
+        """Foo.
```

Comment:
Can these one-liners be examples of using imperative rather than descriptive style, at least if we care about that? E.g.:
`"""Add 5 to the input"""`
as opposed to:
`"""Adds 5 to the input"""`